---
title: "Single-cell exploration of the role of PDAC intratumoral bacteria on tumor and immune cell gene programs"
date: "2024-03-16"
---

```{r setup, include=FALSE}

# Load packages
library(ggplot2) # to output/save plots
library(knitr) # to output/save plots
library(Seurat) # to load Seurat object data

```

  Load somatic cells:
  
```{r pressure, echo=FALSE}

# Load somatic data
USER_FILE_PATH = "/Users/samuelking/Desktop/TotalTissue.combined.harmony_scaled.RData"
steele_somatic_data <- load(USER_FILE_PATH)

```

  From somatic cells, visualize cell clusters using UMAP:

```{r pressure, echo=FALSE}

# Visualize somatic data
somatic_umap <- DimPlot(object = TotalTissue.combined.harmony, reduction = "umap", label = TRUE, pt.size = 0.5) + xlab("UMAP1") + ylab("UMAP2") + theme(axis.line = element_blank(), panel.border = element_blank(), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.background = element_blank(), axis.ticks = element_blank(), axis.text = element_blank())

# Save plot
ggsave("UMAP_Steele_somatic_data.png", somatic_umap, width = 6, height = 4, units = "in", dpi = 300)
somatic_umap

```

  Identify the top genes expressed in each cell cluster:

```{r pressure, echo=FALSE}

annotations <- rownames(TotalTissue.combined.harmony)
current.cluster.ids <- c(0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,
                         29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53)

names(x = current.cluster.ids)
TotalTissue.combined.harmony_Trial <- RenameIdents(object = TotalTissue.combined.harmony, current.cluster.ids)

#Identify gene markers in each cluster

all_markers <-FindAllMarkers(TotalTissue.combined.harmony_Trial, min.pct = 0.25, min.diff.pct = 0.25)

#write results to a csv
write.csv(all_markers, file = "/home/users/ansa1134/all_markers.csv", quote = F)

top_genes <- read.csv("/Users/antoniosalcido/Downloads/all_markers.csv")

#Sorting based off of order of significant genes, and average log fold change. ``
top_genes <- top_genes[order(top_genes$cluster, top_genes$avg_log2FC, decreasing = TRUE), ]

#Keep rows that have an adjusted P value less than 0.05 
top_gene_sig <- top_genes[top_genes$p_val_adj <= 0.05,]

#Now keep just the genes of interest / genes used to identify cell clusters that was reported in the SAHMI paper github. 

genes_keep <- c('PRSS1', 'CTRB2','REG1A','CLU','MKI67','KRT8','SPINK1','KRT19','KRT18', 'KRT18','TFF1','MUC1','ACTA2', 'CDH11', 'PDGFRB', 'COL1A1', 'COL3A1', 
                'RGS5', 'IGFBP7', 'PDPN','DCN','MCAM','IL6','APOE','GLI1','GLI2','GLI3','PDGFA','VWF','CDH5','TPSAB1','CPA3','CD14','ITGAM','FCGR3A','FCGR3B','APOE',
                'C1QA','MARCO','LYZ','HLA-DRA','ITGAE','LYZ','CLEC9A','BATF3','IRF8','IDO1','CD207','CD1A'
                ,'CD1C', 'HLA-DRA','CCL22','LAMP3','IL22RA2','CD101','CD2', 'CD3D','CD3E','NCAM1','NKG7','CD4','CD8A','PRF1'
                ,'IFNG', 'GZMB','CD69','FOXP3','TIGIT','TOP2A','FCGR3A','CD79A', 'MS4A1','CD20','CD138','IGJ','IGLL5','CXCR4','CD117'
                ,'CD27','HLA-DRA')

top_gene_sig <- top_gene_sig[top_gene_sig$gene %in% genes_keep,]

#Classify genes to their categories, based off of what the Steele code did

#Create a function for classification:
cell_types <- function(gene) {
  if (gene %in% c('PRSS1', 'CTRB2',
                  'REG1A','CLU','MKI67','KRT8','SPINK1','KRT19','KRT18', 'KRT18','TFF1','MUC1')) {
    return("Epithelial")
  } else if (gene %in% c('ACTA2', 'CDH11', 'PDGFRB', 'COL1A1', 'COL3A1', 'RGS5', 'IGFBP7', 'PDPN','DCN','MCAM','IL6','APOE','GLI1','GLI2','GLI3','PDGFA')) {
    return("Fibroblasts")
  } else if (gene %in% c('VWF','CDH5')) {
    return("Endothelial")
  } else if (gene %in% c('TPSAB1','CPA3')) {
    return("Mast")
  } else if (gene %in% c('CD14','ITGAM','FCGR3A','FCGR3B','APOE','C1QA','MARCO','LYZ','HLA-DRA')) {
    return("Myeloid")
  } else if (gene %in% c('ITGAE','LYZ','CLEC9A','BATF3','IRF8','IDO1','CD207','CD1A','CD1C', 'HLA-DRA','CCL22','LAMP3','IL22RA2','CD101')) {
    return("DCs")
  } else if (gene %in% c('CD2', 'CD3D','CD3E','NCAM1','NKG7','CD4','CD8A','PRF1','IFNG', 'GZMB','CD69','FOXP3','TIGIT','TOP2A','FCGR3A')) {
    return("T_NK")
  } else {
    return("B")
  }
}

#Add another column to organize genes that are associated with certain cell types 
top_gene_sig$cell_type <- sapply(top_gene_sig$gene, cell_types)

#Save to local machine to view results to manually label the clusters based on the top genes expressed in each cluster
write.csv(top_gene_sig, "/Users/antoniosalcido/Downloads/final_cluster_labeling.csv")

```

  Annotate cell types:

```{r}

new.cluster.ids <- c("CD8+ T cells", "CD4+ T cells", "Myeloid", "Epithelial", "Epithelial", "Myeloid", "Epithelial", "Mast", "Myeloid", "Myeloid", "Myeloid", "CD4+ T cells", "Fibroblasts", "Unknown", "Myeloid", "Myeloid", "B cells", "Epithelial", "Myeloid", "B cells", "Epithelial", "Myeloid", "CD4+ T cells", "Fibroblasts", "Epithelial", "NK cells", "Epithelial", "Epithelial", "Epithelial", "Epithelial", "Epithelial", "Epithelial", "Myeloid", "Epithelial", "Fibroblasts", "Epithelial", "NK cells", "DCs", "CD8+ T cells", "Epithelial", "Endothelial", "B cells", "CD4+ T cells", "Epithelial", "NK cells", "Epithelial", "Epithelial", "Epithelial", "Unknown", "Unknown", "Epithelial", "B cells", "CD8+ T cells", "Epithelial")
names(x = new.cluster.ids) <- levels(x = TotalTissue.combined.harmony)
TotalTissue.combined.harmony <- RenameIdents(object = TotalTissue.combined.harmony, new.cluster.ids)
somatic_umap_celltype <- DimPlot(object = TotalTissue.combined.harmony, reduction = "umap", label = TRUE, pt.size = 0.5) + xlab("UMAP1") + ylab("UMAP2") + theme(axis.line = element_blank(), panel.border = element_blank(), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.background = element_blank(), axis.ticks = element_blank(), axis.text = element_blank())

# Save plot
ggsave("UMAP_Steele_somatic_data_with_cell_types.png", somatic_umap_celltype, width = 6, height = 4, units = "in", dpi = 300)
somatic_umap_celltype

```

### FIGURE 1D ###
  
  QC the somatic data and set a % mitochondrial read threshold of 20% (the standard for cancer cells):
  
```{r pressure, echo=FALSE}

# Add percent mitochondrial read information
TotalTissue.combined.harmony[["percent.mt"]] <- PercentageFeatureSet(TotalTissue.combined.harmony, pattern = "^MT-")

somatic_qc_celltyped <- VlnPlot(TotalTissue.combined.harmony, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3, pt.size=0) + geom_hline(yintercept=20, linetype="dashed", color="black")
ggsave("QC_Steele_somatic_data_with_cell_types_no_dots.png", somatic_qc_celltyped, width = 10, height = 4, units = "in", dpi = 300)
somatic_qc_celltyped

```
  
  Filter out the cells above the % MT threshold:
  
```{r}

TotalTissue.combined.harmony <- subset(TotalTissue.combined.harmony, subset = percent.mt < 20)

somatic_qc_celltyped <- VlnPlot(TotalTissue.combined.harmony, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3, pt.size=0)
ggsave("QC_Steele_somatic_data_with_cell_types_no_dots_mt_removed.png", somatic_qc_celltyped, width = 10, height = 4, units = "in", dpi = 300)
somatic_qc_celltyped

```

### FIGURE 1C ###
 
  Plot the cell type UMAP again:
  
```{r}

somatic_umap_celltype <- DimPlot(object = TotalTissue.combined.harmony, reduction = "umap", label = TRUE, pt.size = 0.5) + xlab("UMAP1") + ylab("UMAP2") + theme(axis.line = element_blank(), panel.border = element_blank(), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.background = element_blank(), axis.ticks = element_blank(), axis.text = element_blank())

# Save plot
ggsave("UMAP_Steele_somatic_data_with_cell_types_mt_removed.png", somatic_umap_celltype, width = 6, height = 4, units = "in", dpi = 300)
somatic_umap_celltype

```

### FIGURE 1E ###

  Show ~top two marker genes for each cell type:

```{r}
# Dot plot
dotplot_top_markers <- DotPlot(TotalTissue.combined.harmony, features = c("CDH5","VWF","IDO1", "CD1C", "NKG7", "PRF1", "IGLL5", "IGJ", "IGFBP7", "COL1A1", "TPSAB1", "CPA3", "KRT19", "MUC1", "HLA-DRA", "LYZ", "CD4", "CD3E", "CD8A", "CD3D"), cols = c('red', 'blue'), dot.scale = 5) + RotatedAxis() + FontSize(x.text = 8, y.text = 8)

# Save plot
ggsave("Dotplot_top_markers.png", dotplot_top_markers, width = 6, height = 4, units = "in", dpi = 300)
dotplot_top_markers
```

### FIGURE 1F ###

  Use inferCNV to analyze the copy number variation of epithelial cells:

```{r}

#devtools::install_github("broadinstitute/infercnv") # Install latest version of infercnv
library(infercnv)

PDAC_TISSUE_7.data <- Read10X("/Users/samuelking/Desktop/GSE155698_RAW/PDAC_TISSUE_7/filtered_feature_bc_matrix")
infercnv_obj = CreateInfercnvObject(
  raw_counts_matrix=PDAC_TISSUE_7.data,
  annotations_file="/Users/samuelking/Downloads/pdac_tissue_7_annotation.txt",
  delim="\t",
  gene_order_file="/Users/samuelking/Downloads/gencode_v19_gene_pos.txt",
  ref_group_names=c("Myeloid",	"NK cells",	"B cells",	"CD4+ T cells",	"Unknown",	"Mast",	"CD8+ T cells",	"DCs",	"Fibroblasts", "Endothelial"))

out_dir = tempfile()
infercnv_obj_default = infercnv::run(
    infercnv_obj,
    cutoff=0.1, # cutoff=1 works well for Smart-seq2, and cutoff=0.1 works well for 10x Genomics
    out_dir=out_dir,
    cluster_by_groups=TRUE, 
    plot_steps=FALSE,
    denoise=TRUE,
    HMM=FALSE,
    no_prelim_plot=TRUE,
    png_res=60
)

# Output the infercnv heatmap to local computer
plot_cnv(infercnv_obj_default, out_dir = ".")

```

  Load microbial data provided by Ghaddar et al. 2022, an analysis of data from Steele et al. 2020:

```{r setup, include=FALSE}

# Load microbiome data; already a Seurat object
USER_FILE_PATH = "/Users/samuelking/Desktop/steele_microbiome.RDS"
steele_microbiome_data <- readRDS(USER_FILE_PATH)

```

### FIGURE 1G ###

  QC of bacterial data: here, number of unique genes (nFeature_RNA) are taxa and number of unique molecules (nCount_RNA) are taxa counts per cell

```{r pressure, echo=FALSE}

steele_microbiome_data[["percent.mt"]] <- PercentageFeatureSet(steele_microbiome_data, pattern = "^MT-")
microbiome_qc <- VlnPlot(steele_microbiome_data, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3, alpha=0.1)

ggsave("QC_Steele_microbiome.png", microbiome_qc, width = 6, height = 4, units = "in", dpi = 300)
microbiome_qc

```
  
  Get barcode lists from somatic and bacterial data:
  
```{r pressure, echo=FALSE}

# Barcodes in somatic cells stored in: TotalTissue.combined.harmony@assays$RNA@cells@dimnames
barcode_list_somatic <- TotalTissue.combined.harmony@assays$RNA@cells@dimnames[[1]] # 43583 barcodes
barcode_list_somatic <- as.matrix(barcode_list_somatic) # make into matrix
  # Remove '-1' in barcodes
barcode_list_somatic_subset <- gsub("-1", "", barcode_list_somatic) # remove -1 label from all somatic cells
barcode_list_somatic_subset <- as.matrix(barcode_list_somatic_subset) # redundant?
  # Remove patient number in barcodes
barcode_list_somatic_nopatient <- gsub("_.*", "", barcode_list_somatic_subset) # 43583 barcodes, remove anything after underscore ("_") in barcode names
  # Save files as CSV
write.csv(barcode_list_somatic_nopatient, file = "barcode list somatic with no patient")

# Barcodes in bacterial cells stored in: 
barcode_list_bacterial <- steele_microbiome_data@assays$RNA@counts@Dimnames[[2]] # 324616 barcodes
barcode_list_bacterial <- as.matrix(barcode_list_bacterial) # make into matrix
  # Remove random number in barcodes
barcode_list_bacterial_nopatient <- gsub("_.*", "", barcode_list_bacterial) # remove anything after an underscore ("_") in barcode names
  # Save file as CSV
write.csv(barcode_list_bacterial, file = "barcode list bacterial")

```

  Overlap barcodes found in somatic and bacterial data, then make a vector that assigns yes/no to somatic cell bacterial association:

```{r}

# Rename barcodes

# Replace barcode suffixes with their corresponding hits from the overlap matrix
new_colnames <- gsub("_9", "-1_10", colnames(steele_microbiome_data))
new_colnames <- gsub("_7", "-1_11", new_colnames)
new_colnames <- gsub("_3", "-1_15", new_colnames) # Note: all downstream "_3" become "-1_15"
new_colnames <- gsub("_5", "-1_16", new_colnames)
new_colnames <- gsub("_23", "-1_17", new_colnames)
new_colnames <- gsub("_24", "-1_17", new_colnames)
new_colnames <- gsub("_27", "-1_1", new_colnames)
new_colnames <- gsub("_21", "-1_3", new_colnames)
new_colnames <- gsub("-1_154", "-1_4", new_colnames)
new_colnames <- gsub("-1_155", "-1_4", new_colnames)
new_colnames <- gsub("-1_156", "-1_4", new_colnames)
new_colnames <- gsub("_18", "-1_5", new_colnames)
new_colnames <- gsub("-1_157", "-1_6", new_colnames)
new_colnames <- gsub("-1_158", "-1_6", new_colnames)
new_colnames <- gsub("-1_159", "-1_6", new_colnames)
new_colnames <- gsub("_19", "-1_7", new_colnames)
new_colnames <- gsub("_16", "-1_8", new_colnames)
new_colnames <- gsub("_12", "-1_9", new_colnames)
new_colnames <- gsub("-1_153", "-1_10", new_colnames)
new_colnames <- gsub("_14", "-1_13", new_colnames)
new_colnames <- gsub("-1_150", "-1_14", new_colnames)
new_colnames <- gsub("-1_151", "-1_14", new_colnames)

#write.csv(new_colnames, file="barcode_sanity_check.csv") # Sanity check for correct barcode suffixes

# Rename cells in taxa data set with updated barcodes
taxa_counts_matrix_renamed <- steele_microbiome_data@assays$RNA@data
colnames(taxa_counts_matrix_renamed) <- new_colnames

# For purposes of assigning the logical microbe-associated/non-associated column in TotalTissue object, collapse groups with redundant cell barcodes
new_colnames_unique <- unique(new_colnames) # 262750 barcodes

# Make taxa counts matrix only have unique colnames
taxa_counts_matrix_renamed <- taxa_counts_matrix_renamed[, new_colnames_unique, drop=FALSE]

# Count overlapping cells between PDAC and taxa datasets
pdac_barcodes <- TotalTissue.combined.harmony@assays$RNA@cells@dimnames[[1]] # 43583 barcodes
pdac_barcodes <- as.matrix(pdac_barcodes) # make into matrix
TotalTissue.combined.harmony@misc$barcodes <- pdac_barcodes

# Find common barcodes
new_colnames_unique <- as.matrix(new_colnames_unique) # make into matrix
overlapping_cells <- intersect(pdac_barcodes, new_colnames_unique)
length(overlapping_cells) # 14515 barcodes overlap

# Make a matrix that assigns all barcodes in somatic data 'yes' if they're in the overlapping barcode list, and 'no' if not
  # Find which barcodes are in overlap list
overlap_logical <- TotalTissue.combined.harmony@misc$barcodes %in% overlapping_cells # generates TRUE/FALSE vector that is length of all somatic barcodes

# Make taxa counts matrix only have cells that overlap with somatic data
taxa_counts_matrix_overlap <- taxa_counts_matrix_renamed[, overlapping_cells, drop=FALSE]
  
```
  
  Make a cell type x taxa counts matrix:
  
```{r}

# Output CSVs (note: the barcodes in the taxa counts csv will output with "-" as "." for some reason.. make sure to replace them back to "-" after)
write.csv(TotalTissue.combined.harmony@active.ident, file = "cellBCs_and_celltypes.csv")
write.csv(taxa_counts_matrix_overlap, file = "cellBCs_and_taxa_counts.csv")

# Matrix conversion done in Python (see celltype_x_taxa_counts_matrix.ipynb)

```
  
  To add a feature to a FeaturePlot, we need to make it either:
  - An Assay feature (e.g. a gene name - "MS4A1")
  - A column name from meta.data (e.g. mitochondrial percentage - "percent.mito")
  - A column name from a DimReduc object corresponding to the cell embedding values (e.g. the PC 1 scores - "PC_1")
  
```{r}

# Add into meta.data
TotalTissue.combined.harmony <- AddMetaData(object = TotalTissue.combined.harmony, metadata = overlap_logical, col.name = "overlap")
  # Added into meta.data because required for FeaturePlot

# Now we can run UMAP with condition: 'if cell has barcode that is in overlapping_cells, make a different color'
overlapping_cells_plot <- FeaturePlot(
  object = TotalTissue.combined.harmony,
  features = "overlap",
  pt.size = 0.3,
  reduction = "umap",
)

ggsave("Overlapping_somatic_and_bacterial_cells.png", overlapping_cells_plot, width = 5, height = 4, units = "in", dpi = 300)
overlapping_cells_plot

```

  Create most abundant taxa per cell:

```{r}

# We need to associate the overlapping barcodes with their taxa counts.
  # Subset the counts x taxa matrix to include only columns from associated bacteria
#taxa_counts_matrix <- steele_microbiome_data@assays$RNA@data # sparse matrix (mostly 0s) with dimension 29 x 324616
#colnames(taxa_counts_matrix) <- barcode_list_bacterial_nopatient # remove dash and numbers from column names
taxa_counts_matrix_overlap <- taxa_counts_matrix_renamed[, overlapping_cells, drop = FALSE]

# Find most abundant taxa per barcode
most_abundant_species <- apply(taxa_counts_matrix_overlap, 2, function(cell_counts) {
    # Find the index of the row with the maximum count in the current cell
    max_index <- which.max(cell_counts)
    # Get the name of the bacterial species corresponding to the maximum count
    max_species <- rownames(taxa_counts_matrix_overlap)[max_index]
    # Return the name of the most abundant species for this cell
    return(max_species)
})
print(most_abundant_species) # only 16k long at the moment so need to make the size of all somatic data

# Make above list the length of somatic data to add
taxa_association_list <- barcode_list_somatic_nopatient # barcode list length of somatic cells
  # If barcode in abundant species list, make value most abundant species, if not label as 'None'
taxa_association_list <- ifelse(TotalTissue.combined.harmony@misc$barcodes %in% overlapping_cells, most_abundant_species, 'None')

# Add most abundant species feature to metadata
TotalTissue.combined.harmony <- AddMetaData(object = TotalTissue.combined.harmony, metadata = taxa_association_list, col.name = "most_abundant_species")

# Run UMAP with taxa as gene features using DimPlot
# For all taxa
DimPlot(
  object =TotalTissue.combined.harmony, 
  group.by = "most_abundant_species",
  reduction = "umap", 
  #cells.highlight = "campylobacter",
  #cols.highlight = "red",
  #cols = "gray",
  pt.size = 0.1, 
  label = T
)

```

### FIGURE 2B-C ###

  We can also plot 1 taxa at a time:

```{r}

# Plot for 1 taxa at a time:

for (taxa in rownames(taxa_counts_matrix)) {
  
  # Extract taxa abundance barcodes
  taxa_cells <- WhichCells(TotalTissue.combined.harmony, most_abundant_species == taxa)
  one_taxa_association_list <- ifelse(TotalTissue.combined.harmony@misc$barcodes %in% overlapping_cells, taxa_cells, 'FALSE')
  one_taxa_association_list <- as.logical(one_taxa_association_list)
  TotalTissue.combined.harmony <- AddMetaData(object = TotalTissue.combined.harmony, metadata = one_taxa_association_list, col.name = "one_taxa_most_abundant")
  
  # Make each taxa plot
  taxa_plot <- FeaturePlot(
  object = TotalTissue.combined.harmony, # Seurat object
  features = "one_taxa_most_abundant",
  pt.size = 0.01,
  reduction = "umap"
)
  
  # Save plot with gene name
  filename <- paste0("Somatic cells with most abundant taxa being_", taxa, ".png")
  ggsave(filename, plot = taxa_plot)
  }

# Campylobacter ONLY as an example
campylobacter <- WhichCells(TotalTissue.combined.harmony, most_abundant_species == "Campylobacter")
campylobacter_association_list <- ifelse(TotalTissue.combined.harmony@misc$barcodes %in% overlapping_cells, campylobacter, 'FALSE')
campylobacter_association_list <- as.logical(campylobacter_association_list)
TotalTissue.combined.harmony <- AddMetaData(object = TotalTissue.combined.harmony, metadata = campylobacter_association_list, col.name = "campylobacter_most_abundant")

# Run UMAP for 1 taxa
campylobacter_plot <- FeaturePlot(
  object = TotalTissue.combined.harmony, # Seurat object
  features = "campylobacter_most_abundant",
  pt.size = 0.01,
  reduction = "umap"
)

ggsave("Campylobacter_plot.png", campylobacter_plot, width = 5, height = 4, units = "in", dpi = 300)
campylobacter

```

### FIGURE 3A-C ###

  Analyze DEGs in each cell type identified above and use ReactomePA to determine the reactome pathways prominent in each cell cluster:

```{r}

#Change to get the cell types that you are interested in
cell_type = c("CD4+ T cells")

#Subset just the  cell_type of focus from the Overall Seurat Object 
cell_type_Seurat <- subset(TotalTissue.combined.harmony, idents = cell_type)

#Pull out the meta.data column -  which cells are associated with bacteria
cell_type_Bacterial_Cells <- cell_type_Seurat@meta.data$overlap 

Idents(cell_type_Seurat) <- cell_type_Bacterial_Cells

#Perform DEG analysis 

DEG <- FindMarkers(object = cell_type_Seurat,
                   ident.1 = "TRUE",
                   ident.2 = "FALSE",
                   verbose = TRUE)

#Sorting based off of  average log fold change. ``
DEG <- DEG[order(DEG$avg_log2FC, decreasing = TRUE), ]

#Keep rows that have an adjusted P value less than 0.05 
DEG <- DEG[DEG$p_val_adj <= 0.05,]

#Keep rows that have a Pct.1 or Pct.2 > 0.7. This is the threshold that is approximately what the Sahmi paper did

DEG <- DEG[DEG$pct.1 >= 0.7 | DEG$pct.2 >= 0.7,]

#Keep rows that have an avg_log2FC of at least 0.3 - threshold roughly based on what the SAHMI paper did

DEG <- DEG[DEG$avg_log2FC >= 0.3 | DEG$avg_log2FC <= -0.3,]

#Lets just plot the top 10 differential expressed genes, and the bottom 10 expressed genes

top_20 <- head(DEG[order(DEG$avg_log2FC, decreasing = TRUE), ], 10)
bottom_20 <- head(DEG[order(DEG$avg_log2FC), ], 10)
bottom_20 <- bottom_20[order(-bottom_20$avg_log2FC), , drop = FALSE]

DEG_plot <- rbind(top_20, bottom_20)

#DEG_plot <- DEG

DEG_plot$gene_names <- factor(rownames(DEG_plot), levels = rev(rownames(DEG_plot)))
DEG_plot$Regulation <- ifelse(DEG_plot$avg_log2FC > 0, "Upregulated", "Downregulated")
#Use ggplot to create the horizontal bar graphs

ggplot(DEG_plot, aes(x = avg_log2FC, y = gene_names, fill=Regulation)) + 
  geom_bar(stat = "identity") +
  labs(x = "average log2(Fold Change)", y = "") +
  scale_fill_manual(values = c("Upregulated" = "forestgreen", "Downregulated" = "firebrick")) +
  #geom_text(aes(label = gene_names), hjust = 1.5, position = position_dodge(width = 0.9), color = "black", size = 3) +
  ggtitle("CD4+ T cells differentially expressed genes ")

#Perform the DEG pathway Enrichment analysis over representation using Reactome PA - method used in the SAHMI paper

#Take out genes from the DEG_plot structure and turn them into a list

path_genes_bacterial <- as.list(rownames(DEG_plot[DEG_plot$avg_log2FC > 0,]))

path_genes_unassociated <- as.list(rownames(DEG_plot[DEG_plot$avg_log2FC < 0,]))

path_genes_bacterial_trans = bitr(path_genes_bacterial, fromType = "SYMBOL", toType="ENTREZID", OrgDb = "org.Hs.eg.db", drop = "TRUE")
Entrez_Ids_bacterial <- as.list(path_genes_bacterial_trans$ENTREZID)

path_genes_unassociated_trans = bitr(path_genes_unassociated, fromType = "SYMBOL", toType="ENTREZID", OrgDb = "org.Hs.eg.db", drop = "TRUE")
Entrez_Ids_unassociated <- as.list(path_genes_unassociated_trans$ENTREZID)

#Perform Analysis
pathway_enrich_bacterial <- enrichPathway(gene = Entrez_Ids_bacterial, organism = "human", readable = TRUE)
pathway_enrich_unassociated <- enrichPathway(gene = Entrez_Ids_unassociated, organism = "human", readable = TRUE)

```

### FIGURE 3D-F ###

  Create the reactome pathway dot plots:

```{r dot, echo = FALSE}

#Create Dot Plots based on the Reactome Pathway results that you got from the Reactome Pathway

Reactome <- read.csv("/Users/antoniosalcido/Desktop/DEG_Plots/reactome_corrected.csv")

Reactome$logP <- -log(Reactome$P.value)

#Subset Adaptive Immune Cell types

Reactome_Adap <- subset(Reactome, Cell.type %in% c("CD8+ bacterial associated",
                                                   "CD8+ unassociated",
                                                   "CD4+ bacterial associated",
                                                   "CD4+ unassociated",
                                                   "B cells bacterial associated",
                                                   "B cells unassociated"))
#Subset Innate Immune Cell types

Reactome_Inna <- subset(Reactome, Cell.type %in% c("Myeloid bacterial associated",
                                                   "Myeloid unassociated",
                                                   "Mast cells bacterial associated",
                                                   "Mast cells unassociated",
                                                   "NK cells bacterial associated",
                                                   "NK cells unassociated",
                                                   "DCs bacterial associated",
                                                   "DCs unassociated"))
#Subset the resto of Cell types 

Reactome_struct <- subset(Reactome, Cell.type %in% c("Fibroblast bacterial associated",
                                                     "Fibroblast unassociated",
                                                     "Epithelial bacterial associated",
                                                     "Epithelial unassociated",
                                                     "Endothelial cells bacterial associated",
                                                     "Endothelial cells unassociated"))

#Creat dot plot based on the subset of cell types

dot <- ggplot(Reactome_Inna, aes(x = Cell.type, y = reorder(Pathway,logP), size = logP, color = Cell.type)) +
  geom_point() +
  scale_size_continuous(trans = "log10", range = c(1, 10)) +  # Scale dot size using log10 transformation
  labs(x = "Cell Type and Bacterial Association", y = "Reactome Pathways", title = "Overrepresented Reactome Pathways",
       size = "-log(p)", color = "Cell Type") +  # Label the axes
  theme_minimal() +
  #theme_set(theme_gray(base_size = 10, base_family = "Arial", dpi = 300)) +
  theme(axis.text.x = element_text(angle = 20, hjust = 1, color = "black"),
        axis.text.y = element_text(color = "black"),
        axis.title = element_text(face = "bold", color = "black"),
        plot.title = element_text(face = "bold", color = "black"))

ggsave("/Users/antoniosalcido/Downloads/Innate_reactome.png", plot = dot, width = 11 , height = 9, dpi = 2000)

```